File: assignment/week7/rod_cutting.py
def rod_cutting(length_rod, price_i):
    price_i = [0] + price_i
    optimal_price = price_i[:]
    optimal_divide = [0] * (length_rod + 1)
    for i in range(1, length_rod + 1):
        best = price_i[i]
        optimal_divide[i] = i
        for j in range(1, i//2 + 1):
            price = price_i[j] + price_i[i - j]
            if best < price:
                best = price
                optimal_divide[i] = j

        optimal_price[i] = best
    print(optimal_price[length_rod])
    while(length_rod > 0):
        divided_amount = optimal_divide[length_rod]
        print(divided_amount, end=' ')
        length_rod -= divided_amount
    print()
    return optimal_price[length_rod]

rod_cutting(10, [1, 5, 8, 9, 10, 17, 17, 20, 24, 30])
rod_cutting(4, [1,5,8,9])

File: assignment/week7/rod_cutting.cpp
//2015004075_김태형_11802
#include<stdio.h>
#include<stdlib.h>

void reverse_print(int *optimal_divide_amout, int length_rod){
    if (length_rod <= 0) return;
    int i = optimal_divide_amout[length_rod];
    printf("%d ", i);
    reverse_print(optimal_divide_amout, length_rod - i);
}

int main(){
    int length_rod, i, j, price, optimal_price;
    int *price_i;
    int *optimal_price_i;
    int *optimal_divide_amout;
    //input session
    scanf("%d", &length_rod);
    price_i = (int*)calloc(length_rod + 1, sizeof(int));
    optimal_price_i = (int*)calloc(length_rod + 1, sizeof(int));
    optimal_divide_amout = (int*)calloc(length_rod + 1, sizeof(int));
    for(i = 0; i < length_rod; i++){
        scanf("%d", &price);
        price_i[i + 1] = price;
        optimal_price_i[i + 1] = price;
        optimal_divide_amout[i + 1] = i + 1; 
    }
    for(i = 1; i <= length_rod; i++){
        optimal_price = price_i[i];
        for (j = 1; j <= i; j++){
            price = optimal_price_i[j] + optimal_price_i[i - j];
            if (optimal_price < price){
                optimal_price = price;
                optimal_divide_amout[i] = j;
            }
        }
        optimal_price_i[i] = optimal_price;
    }
    printf("%d\n", optimal_price_i[length_rod]);
    reverse_print(optimal_divide_amout, length_rod);
    printf("\n");
    return 0;
}